### DISCUSSION TITLE
No `Provides-Extra: all` in 0.12.0 METADATA

### DISCUSSION THREAD

**Discussion by pjonsson** [fastapi/typer#785] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# With 0.11.1:
$ python3 -m pip download --src :all: typer==0.11.1
$ unzip typer-0.11.1-py3-none-any.whl
$ grep Extra typer-0.11.1.dist-info/METADATA
Provides-Extra: all
$

# Same thing for 0.12.0:
$ python3 -m pip download --src :all: typer==0.12.0
$ unzip typer-0.12.0-py3-none-any.whl
$ grep Extra typer-0.12.0.dist-info/METADATA
$
```

### Description

I started getting the warning: `WARNING: typer 0.12.0 does not provide the extra 'all'` after typer 0.12.0 was released.

#780 seems to preserve the `all` extra for the typer package, so not sure why I suddenly get the warning.

### Operating System

Linux

### Typer Version

>=0.9.4

### Python Version

3.10

**Comment by tiangolo** [comment, marked as answer]
Hey there! You no longer need to install `typer[all]`, now just `typer` includes everything.

You can read more about it in the release notes: https://typer.tiangolo.com/release-notes/#0121

Copied from there:

---

In version `0.12.0`, the `typer` package depends on `typer-slim[standard]` which includes the default dependencies (instead of `typer[all]`) and `typer-cli` (that provides the `typer` command).

If you don't want the extra optional dependencies (`rich` and `shellingham`), you can install `typer-slim` instead.

You can also install `typer-slim[standard]`, which includes the default optional dependencies, but not the `typer` command.

In version `0.12.0` the `typer-cli` package only provides the `typer` command, but the code is still in the main code, so even without installing `typer-cli`, it can be called with `python -m typer`.

This approach of having `typer` depend on `typer-slim[standard]` instead of including the whole code and dependencies itself caused an issue when upgrading from `typer <0.12.0` to `typer >=0.12.0`, see issue [#790](https://github.com/tiangolo/typer/issues/790). This is fixed in version `0.12.1`.

**Reply by pjonsson** [reply]
Hi, sorry for being unclear with my question. I know what changed, I wished typer had kept the `all` option for a while so it's possible to depend on `typer[all] >0.9`, which I thought was the intention with this line: https://github.com/tiangolo/typer/blob/ef9beaf521047d44b0588f9f6927aa128d6b3f72/typer_cli_package/pyproject.toml#L45

**Reply by tiangolo** [reply]
Yep, that was the intention, but PDM removes the `all` optional dependency from the final built artifact (sdist or wheel), and I'm actually not sure if it's possible to have an empty option group. :thinking: 

So, I ended up just removing it from `pyproject.toml` as it was not really affecting anything.

**Reply by gotmax23** [reply]
> Yep, that was the intention, but PDM removes the all optional dependency from the final built artifact (sdist or wheel), and I'm actually not sure if it's possible to have an empty option group. 🤔

It should be possible, but it turns out, there's a bug in the pyproject-metadata library (see https://github.com/pypa/pyproject-metadata/pull/95). Would you be willing to add an `all` extra with a dependency that the main project already depends on until this is fixed?

**Reply by joanise** [reply]
+1: this issue is breaking our projects that use `typer[all]>=0.9.0`. We've solved it by changing that to 
```
typer = ">=0.9.0"
rich = ">=10.11.0"
shellingham = ">=1.3.0"
```
but it would be nicer if the next typer patch defined `[all]` in one way or another, even though it's redundant.
